File: model/ForeignKey.vb
Public Class ForeignKey
    Public Table As Table
    Public Name As String
    Public Columns As New List(Of String)
    Public RefTable As Table
    Public RefColumns As New List(Of String)
    Public OnUpdate As String
    Public OnDelete As String

    Public Sub New(ByVal name As String)
        Me.Name = name
    End Sub

    Public Sub New(ByVal table As Table, ByVal name As String, _
                   ByVal columns As String, ByVal refTable As Table, ByVal refColumns As String)
        Me.New(table, name, columns, refTable, refColumns, "", "")
    End Sub

    Public Sub New(ByVal table As Table, ByVal name As String, _
                   ByVal columns As String, ByVal refTable As Table, ByVal refColumns As String, _
                   ByVal onUpdate As String, ByVal onDelete As String)
        Me.Table = table
        Me.Name = name
        Me.Columns = New List(Of String)(columns.Split(","c))
        Me.RefTable = refTable
        Me.RefColumns = New List(Of String)(refColumns.Split(","c))
        Me.OnUpdate = onUpdate
        Me.OnDelete = onDelete
    End Sub

    Public Function Script() As String
        Dim text As New StringBuilder()
        text.AppendFormat("ALTER TABLE [{0}].[{1}] WITH CHECK ADD CONSTRAINT [{2}]{3}", _
                          Table.Owner, Table.Name, Name, vbCrLf)
        text.AppendFormat("FOREIGN KEY([{0}]) REFERENCES [{1}].[{2}] ([{3}]){4}", _
                          String.Join("], [", Columns.ToArray()), RefTable.Owner, RefTable.Name, _
                          String.Join("], [", RefColumns.ToArray()), vbCrLf)
        Return text.ToString()
    End Function
End Class

File: model/ForeignKey.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace model
{
	public class ForeignKey
	{
		public bool Check;
		public List<string> Columns = new List<string>();
		public string Name;
		public string OnDelete;
		public string OnUpdate;
		public List<string> RefColumns = new List<string>();
		public Table RefTable;
		public Table Table;

		public ForeignKey(string name)
		{
			this.Name = name;
		}

		public ForeignKey(Table table, string name, string columns, Table refTable, string refColumns)
			: this(table, name, columns, refTable, refColumns, "", "")
		{
		}

		public ForeignKey(Table table, string name, string columns, Table refTable, string refColumns, string onUpdate,
			string onDelete)
		{
			this.Table = table;
			this.Name = name;
			this.Columns = new List<string>(columns.Split(','));
			this.RefTable = refTable;
			this.RefColumns = new List<string>(refColumns.Split(','));
			this.OnUpdate = onUpdate;
			this.OnDelete = onDelete;
		}

		public string CheckText
		{
			get { return this.Check ? "CHECK" : "NOCHECK"; }
		}

		private void AssertArgNotNull(object arg, string argName)
		{
			if (arg == null)
			{
				throw new ArgumentNullException(string.Format(
					"Unable to Script FK {0}. {1} must not be null.", this.Name, argName));
			}
		}

		public string ScriptCreate()
		{
			this.AssertArgNotNull(this.Table, "Table");
			this.AssertArgNotNull(this.Columns, "Columns");
			this.AssertArgNotNull(this.RefTable, "RefTable");
			this.AssertArgNotNull(this.RefColumns, "RefColumns");

			var text = new StringBuilder();
			text.AppendFormat("ALTER TABLE [{0}].[{1}] WITH {2} ADD CONSTRAINT [{3}]\r\n", this.Table.Owner, this.Table.Name, this.CheckText, this.Name);
			text.AppendFormat("   FOREIGN KEY([{0}]) REFERENCES [{1}].[{2}] ([{3}])\r\n", string.Join("], [", this.Columns.ToArray()), this.RefTable.Owner, this.RefTable.Name, string.Join("], [", this.RefColumns.ToArray()));
			if (!string.IsNullOrEmpty(this.OnUpdate))
			{
				text.AppendFormat("   ON UPDATE {0}\r\n", this.OnUpdate);
			}
			if (!string.IsNullOrEmpty(this.OnDelete))
			{
				text.AppendFormat("   ON DELETE {0}\r\n", this.OnDelete);
			}
			if (!this.Check)
			{
				text.AppendFormat("   ALTER TABLE [{0}].[{1}] NOCHECK CONSTRAINT [{2}]\r\n", this.Table.Owner, this.Table.Name, this.Name);
			}
			return text.ToString();
		}

		public string ScriptDrop()
		{
			return string.Format("ALTER TABLE [{0}].[{1}] DROP CONSTRAINT [{2}]\r\n", this.Table.Owner, this.Table.Name, this.Name);
		}
	}
}